### PR TITLE
Prefer statx on linux if available

### LINKED ISSUE #59743: `Metadata::created()` not working on Linux with `statx`
First of all, it now returns an Err with "not available on this platform currently".
But I found that there is a system call [`statx`](http://man7.org/linux/man-pages/man2/statx.2.html) available in Linux kernel 4.11, which can get the birth time. 

As the [doc](https://doc.rust-lang.org/nightly/std/fs/struct.Metadata.html#method.created), the function uses  `stat` on Unix platforms. But the [`stat` result](https://docs.rs/libc/0.2.51/libc/struct.stat.html) just never contains `birthtime`.

Shouldn't it try `statx` when available?

### PR DESCRIPTION
**Pull request by oxalica**
This PR make `metadata`-related functions try to invoke `statx` first on Linux if available,
making `std::fs::Metadata::created` work on Linux with `statx` supported.

It follows the discussion in #61386 , and will fix #59743 

The implementation of this PR is simply converting `struct statx` into `struct stat64` with
extra fields for `btime` if `statx` succeeds, since other fields are not currently used.

---

I also did a separated benchmark for `fs::metadata`, `stat64`, `statx`, and `statx` with conversion to `stat64`.
It shows that `statx` with conversion is even more faster than pure `statx`.
I think it's due to `sizeof stat64 == 114` but `sizeof statx == 256`.

Anyway, the bare implementation of `statx` with conversion is only about 0.2% slower than the original impl (`stat64`-family).
With heap-allocation counted (~8.5% of total cost), the difference between `stat` and `statx` (with or without conversion) is just nothing.

Therefore, I think it is not urgent to use bare `struct statx` as underlying representation now.
There is no need to break `std::os::linux::fs::MetadataExt::as_raw_stat` (https://github.com/rust-lang/rust/pull/61386#issuecomment-538074018)

[Separated bare benchmarks](https://gist.github.com/oxalica/c4073ecb202c599fe41b7f15f86dc79c):
```
metadata_ok             time:   [529.41 ns 529.77 ns 530.19 ns]
metadata_err            time:   [538.71 ns 539.39 ns 540.35 ns]
stat64_ok               time:   [484.32 ns 484.53 ns 484.75 ns]
stat64_err              time:   [481.77 ns 482.00 ns 482.24 ns]
statx_ok                time:   [488.07 ns 488.35 ns 488.62 ns]
statx_err               time:   [487.74 ns 488.00 ns 488.27 ns]
statx_cvt_ok            time:   [485.05 ns 485.28 ns 485.53 ns]
statx_cvt_err           time:   [485.23 ns 485.45 ns 485.67 ns]
```

r? @alexcrichton

### CODE REVIEW

**Comment by ariasuni** [comment]
I think the doc needs to be updated, see [here](https://doc.rust-lang.org/std/fs/struct.Metadata.html#method.created):

> The returned value corresponds to the `birthtime` field of stat on Unix platforms and the `ftCreationTime` field on Windows platforms.

It could be changed to something like:

> The returned value corresponds to the `btime` field of statx on Linux (available since Linux 4.11), `birthtime` field of stat on other Unix platforms and the `ftCreationTime` field on Windows platforms.

**Comment by alexcrichton** [comment]
This is looking good to me, thanks! My main comment would be to make the platform-specific code here a bit more maintainable. Would it be possible to instead of duplicating each function only include Linux-specific code like so:

```rust
#[cfg(target_os = "linux")]
{
    // statx code...
}

// the current implementation as a fallback
```

That should hopefully make it a bit more clear that this is the same as the current code, only it has a Linux-specific fallback.

File: src/libstd/sys/unix/fs.rs [hunk 1]
```diff
@@ -259,12 +248,28 @@ impl FileAttr {
...
+            if let Some(ext) = &self.statx_extra_fields {
+                return if (ext.stx_mask & libc::STATX_BTIME) != 0 {
+                    Ok(SystemTime::from(libc::timespec {
+                        tv_sec: ext.stx_btime.tv_sec as libc::time_t,
+                        tv_nsec: ext.stx_btime.tv_nsec as libc::c_long,
+                    }))
+                } else {
+                    Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        "creation time is not available for the filesystam",
```
Comment (ariasuni) [inline src/libstd/sys/unix/fs.rs]: There’s a typo: filesystam → filesystem

File: src/libstd/sys/unix/fs.rs [hunk 2]
```diff
@@ -44,6 +48,83 @@ pub struct File(FileDesc);
...
+            stat.st_blocks = buf.stx_blocks as i64;
+            stat.st_atime = buf.stx_atime.tv_sec;
+            stat.st_atime_nsec = buf.stx_atime.tv_nsec as i64;
+            stat.st_mtime = buf.stx_mtime.tv_sec;
+            stat.st_mtime_nsec = buf.stx_mtime.tv_nsec as i64;
+            stat.st_ctime = buf.stx_ctime.tv_sec;
+            stat.st_ctime_nsec = buf.stx_ctime.tv_nsec as i64;
+
+            let extra = StatxExtraFields {
+                stx_mask: buf.stx_mask,
```
Comment (alexcrichton) [inline src/libstd/sys/unix/fs.rs]: This `stx_mask` field, according to the man page, looks to indicate whether fields are actually available or not. This is used for the creation time, but should this also be taken into account for all the fields above as well? (presumably for them we'd return an error if they aren't available?)

File: src/libstd/sys/unix/fs.rs [hunk 3]
```diff
@@ -20,6 +20,10 @@ use libc::{stat64, fstat64, lstat64, off64_t, ftruncate64, lseek64, dirent64, re
 use libc::fstatat64;
 #[cfg(any(target_os = "linux", target_os = "emscripten", target_os = "android"))]
 use libc::dirfd;
+// We only use struct `statx`, not the function `statx`.
+// Instead, use `syscall` to check if it is available at runtime.
+#[cfg(target_os = "linux")]
+use libc::{statx, makedev};
```
Comment (alexcrichton) [inline src/libstd/sys/unix/fs.rs]: Since this is only used in the `try_statx` function, could this `use` be moved down there?

**Review by alexcrichton — COMMENTED** [review]
Just a few final comments, but otherwise r=me, thanks!

File: src/libstd/sys/unix/fs.rs (hunk 2 above)
Comment (oxalica) [reply src/libstd/sys/unix/fs.rs]: So we need to check `buf.statx_mask & STATX_BASIC_STATS == STATX_BASIC_STATS`?

But what would be returned by `stat64` for the case of missing basic fields?
I haven't seen anything about it in the man page of `stat`, including the `ERRORS` section.

For [man page of `statx`](http://man7.org/linux/man-pages/man2/statx.2.html),
> If a filesystem does not support a field or if it has an unrepre‐
> sentable value (for instance, a file with an exotic type), then the
> mask bit corresponding to that field will be cleared in stx_mask even
> if the user asked for it **and a dummy value will be filled in for com‐**
> **patibility purposes if one is available** (e.g., a dummy UID and GID
> may be specified to mount under some circumstances).

Is this the same behavior as `stat` which never return a "partial result"?

BTW: I tested to `statx` a file in mounted FAT32 filesystem which has no permission support. The result has all `STATX_BASIC_STATS` set in `stx_mode`, and the file mode, uid and gid are just what specified in `mount`. And `stat64` also returns the dummy result without any error.

File: src/libstd/sys/unix/fs.rs (hunk 3 above)
Comment (oxalica) [reply src/libstd/sys/unix/fs.rs]: It will collide `fn statx` in the function body.
I think we can simply remove the `use` and refer them with `libc::`.

File: src/libstd/sys/unix/fs.rs (hunk 2 above)
Comment (alexcrichton) [reply src/libstd/sys/unix/fs.rs]: Ah ok, this sounds reasonable to me, thanks for the clarification!

**Comment by oxalica** [comment]
Oh, sorry, my fault. A cfg guard is missing.

**Comment by oxalica** [comment]
I took a look at `libc`.

The `struct statx` and related flags is supported for `cfg(all(target_os = "linux", target_env = "gnu"))`.
But `SYS_statx` is not always with them. It is only available on arch `x86`, `arm`, `powerpc`, `x86_64`, `powerpc64`, `sparc64` (also, requires linux-gnu), and missing on arch `mips`, `aarch64`, `mips64`, `s390x`.

I'm not sure if we should check all `target_os`, `target_env` and `target_arch` for the related codes, which is ugly and long.
Would it be better to use a macro instead of `cfg` block everywhere?


```rust
#[cfg(all(target_os = "linux", target_env = "gnu", any(
    target_arch = "x86",
    target_arch = "arm",
    target_arch = "powerpc",
    target_arch = "x86_64",
    target_arch = "powerpc64",
    target_arch = "sparc64",
)))]
```

@alexcrichton

**Comment by mati865** [comment]
FWIW statx should also work with musl but `libc` crate doesn't expose it. It definitely should *not* block this PR though.

**Comment by oxalica** [comment]
@mati865 

Do you mean we should hand-write the struct and relative constants? 
The main issue is that `SYS_statx` seems to be different between platforms.
And I also have no device to test these platforms.

**Comment by oxalica** [comment]
Some discussion about musl syscall ids of libc can be found in https://github.com/rust-lang/libc/pull/699#issuecomment-318380576 
The syscalls `SYS_pkey_*` have been removed because of the minimum version support of musl. `SYS_statx` is added **even later than them**, and requires later version.

Maybe adding them to `libc` can be difficult.

**Comment by mati865** [comment]
@oxalica a lot has changed in `libc` since that comment. As I said you should *not* block on musl.
Once musl symbols are added to `libc` statx can be enabled for it but to not delay things further making it available only to gnu (as tier1 platform) should be fine (however I'm not decisive here).

**Comment by oxalica** [comment]
@mati865 

The main point is not `musl` but we still need lots of conditions on `target_arch` on `gnu`, since `SYS_statx` is not available on all architecture in `libc`.

> ```
> #[cfg(all(target_os = "linux", target_env = "gnu", any(
>     target_arch = "x86",
>     target_arch = "arm",
>     target_arch = "powerpc",
>     target_arch = "x86_64",
>     target_arch = "powerpc64",
>     target_arch = "sparc64",
> )))]
> ```

Now I have cfg blocks in every call site. Replacing all cfgs will make it redundant and have poor maintainability.

**Comment by oxalica** [comment]
(Rebased)

I write a large cfg to only enable it only on what `libc` is currently supported.
I think macro can make it easier to maintain when `libc` support `statx` structs and constants on more arch or musl.

Hope it compile. (It's quite hard to test all architectures manually)

@alexcrichton

**Comment by Centril** [comment]
Failed in https://github.com/rust-lang/rust/pull/65449#issuecomment-542399088, @bors r-

**Comment by oxalica** [comment]
@alexcrichton Fixed.

File: src/libstd/fs.rs [hunk 4]
```diff
@@ -1090,13 +1090,14 @@ impl Metadata {
 
     /// Returns the creation time listed in this metadata.
     ///
-    /// The returned value corresponds to the `birthtime` field of `stat` on
+    /// The returned value corresponds to the `btime` field of `statx` on
+    /// Linux not prior to 4.11, the `birthtime` field of `stat` on other
```
Comment (RalfJung) [inline src/libstd/fs.rs]: "not prior" is hard to parse; what about "starting from" or so?

You are basically saying `!(linux_version < (4, 11))` instead of the much more comprehensible `linux_version >= (4, 11)`.

Comment (oxalica) [reply src/libstd/fs.rs]: Fixed

**Comment by Centril** [comment]
Failed in https://github.com/rust-lang/rust/pull/65530#issuecomment-543417994, @bors r- rollup=never

**Comment by oxalica** [comment]
@alexcrichton 
Squashed previous commits and rebased to master.
Fixed cast of `stx_btime.tv_nsec`.

**Comment by jethrogb** [comment]
I think this might be broken on some filesystems on Linux. Simple `fs::exists` calls stopped working for me when using Docker.

**Comment by jethrogb** [comment]
Actually, I think docker may have a cgroup or BPF filter or so installed that returns EPERM on non-whitelisted syscalls. This prevents the fallback from operating, because that only triggers on ENOSYS.

Host kernel: 4.15.0-65-generic

docker version
```
Client:
 Version:      17.03.2-ce
 API version:  1.27
 Go version:   go1.6.2
 Git commit:   f5ec1e2
 Built:        Thu Jul  5 23:07:48 2018
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.2-ce
 API version:  1.27 (minimum version 1.12)
 Go version:   go1.6.2
 Git commit:   f5ec1e2
 Built:        Thu Jul  5 23:07:48 2018
 OS/Arch:      linux/amd64
 Experimental: false
```

```
$ docker run --rm -it ubuntu:xenial
# apt update
# apt install wget
# wget https://sh.rustup.rs -O rustup.sh
# chmod ugo+x rustup.sh 
# ./rustup.sh -y --default-toolchain nightly
# source $HOME/.cargo/env
# rustc -vV
rustc 1.40.0-nightly (7979016af 2019-10-20)
binary: rustc
commit-hash: 7979016aff545f7b41cc517031026020b340989d
commit-date: 2019-10-20
host: x86_64-unknown-linux-gnu
release: 1.40.0-nightly
error: failed to find a `codegen-backends` folder in the sysroot candidates:
* /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu
* /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu
```
